src/borgstore/backends/rclone.py

```diff
@@ -0,0 +1,300 @@
+
+    def _requests(self, fn, *args, **kwargs):
+        """
+        Runs a call to the requests function fn with *args and **kwargs
+
+        It adds auth and decodes errors in a consistent way
+
+        It returns the response object
+        """
+        if not self.process or not self.url:
+            raise BackendMustBeOpen()
+        r = fn(*args, auth=(self.user, self.password), **kwargs)
+        if r.status_code == 404:
+            raise ObjectNotFound(f"Not Found: error {r.status_code}: {r.text}")
+        elif r.status_code not in (200, 206):
+            raise BackendError(f"rclone rc command failed: error {r.status_code}: {r.text}")
+        return r
+
+    def _rpc(self, command, json_input, **kwargs):
+        """
+        Run the rclone command over the rclone API
+
+        Additional kwargs may be passed to requests
+        """
+        if not self.url:
+            raise BackendMustBeOpen()
+        r = self._requests(
+            requests.post,
+            self.url+command,
+            json=json_input,
+            **kwargs,
+        )
+        return r.json()
+
+    def create(self):
+        """create (initialize) the rclone storage"""
+        if self.process:
+            raise BackendMustNotBeOpen()
+        with self:
+            info = self.info("")
+            if info.exists:
+                raise BackendAlreadyExists(f"rclone storage base path already exists: {self.fs}")
+            self.mkdir("")
+
+    def destroy(self):
+        """completely remove the rclone storage (and its contents)"""
+        if self.process:
+            raise BackendMustNotBeOpen()
+        with self:
+            info = self.info("")
+            if not info.exists:
+                raise BackendDoesNotExist(f"rclone storage base path does not exist: {self.fs}")
+            self._rpc("operations/purge", {
+                "fs": self.fs,
+                "remote": "",
+            })
+
+    def __enter__(self):
+        self.open()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+        return False
+
+    def mkdir(self, name: str) -> None:
+        """create directory/namespace <name>"""
+        validate_name(name)
+        self._rpc("operations/mkdir", {
+            "fs": self.fs,
+            "remote": name,
+        })
+
+    def rmdir(self, name: str) -> None:
+        """remove directory/namespace <name>"""
+        validate_name(name)
+        self._rpc("operations/rmdir", {
+            "fs": self.fs,
+            "remote": name,
+        })
+
+    def _to_item_info(self, remote, item, dir_size=0):
+        """Converts an rclone item at remote into a borgstore ItemInfo
+        """
+        if item is None:
+            return ItemInfo(name=os.path.basename(remote), exists=False, directory=False, size=0)
+        name = item["Name"]
+        size = item["Size"]
+        directory = item["IsDir"]
+        if directory:
+            size = dir_size
```

Comment:
If you are sure you have the real size of the directory, please return it.

`Store.list` can then do less `backend.list` calls in the recursion if it sees beforehand the directory is empty.

Related tests and maybe other backends need fixing.

Reply:
If you mean size of the directory as "number of items inside it" then rclone doesn't know that. A small number of the backends support that, but in general we don't know. None of the major backends support it :-(

Rclone uses size of the directory as size of the directory on disk, 

```
$ ls -ld /
drwxr-xr-x 21 root root 4096 Apr 15  2022 /
```

So would normally output 4096

Reply:
Ah, interesting.

On macOS size == entries * 32.

I'll do a separate PR now fixing that stuff. We just can't use that, either it is OS (fs?) specific or not even known.

Reply:
#48 / master